src/Form/DataTransformerResolver.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Form;
+
+use Sonata\AdminBundle\Admin\FieldDescriptionInterface;
+use Sonata\AdminBundle\Form\DataTransformer\BooleanToStringTransformer;
+use Sonata\AdminBundle\Form\DataTransformer\ModelToIdTransformer;
+use Sonata\AdminBundle\Model\ModelManagerInterface;
+use Symfony\Component\Form\DataTransformerInterface;
+use Symfony\Component\Form\Extension\Core\DataTransformer\DateTimeToStringTransformer;
+
+/**
+ * @author Peter Gribanov <info@peter-gribanov.ru>
+ */
+final class DataTransformerResolver
+{
+    /**
+     * @var array<string, DataTransformerInterface>
+     */
+    private $globalCustomTransformers = [];
+
+    /**
+     * @param array<string, DataTransformerInterface> $customGlobalTransformers
+     */
+    public function __construct(array $customGlobalTransformers = [])
+    {
+        foreach ($customGlobalTransformers as $fieldType => $dataTransformer) {
+            $this->addCustomGlobalTransformer($fieldType, $dataTransformer);
+        }
+    }
+
+    public function addCustomGlobalTransformer(string $fieldType, DataTransformerInterface $dataTransformer): void
+    {
+        $this->globalCustomTransformers[$fieldType] = $dataTransformer;
+    }
+
+    public function resolve(
+        FieldDescriptionInterface $fieldDescription,
+        ModelManagerInterface $modelManager
+    ): ?DataTransformerInterface {
+        $dataTransformer = $fieldDescription->getOption('data_transformer');
+
+        // allow override predefined transformers for 'date', 'boolean' and 'choice' field types
+        if ($dataTransformer instanceof DataTransformerInterface) {
+            return $dataTransformer;
+        }
+
+        $fieldType = (string) $fieldDescription->getType();
+
+        // allow override predefined transformers on a global level
+        if (\array_key_exists($fieldType, $this->globalCustomTransformers)) {
+            return $this->globalCustomTransformers[$fieldType];
+        }
+
+        // Handle date type has setter expect a DateTime object
+        if ('date' === $fieldDescription->getType()) {
+            return new DateTimeToStringTransformer(null, null, 'Y-m-d');
```

Comment:
Can you register this transformer as services, so you do not create new instances every time.

Reply:
Symfony has the principle that Data Transformers are initialized each time they are used. Data Transformers are not registered as services.
https://symfony.com/doc/current/form/data_transformers.html

I add lazy-load the predefined data transformers.